File: internal/export/sarif.go
package export

import (
	"encoding/json"

	"github.com/sentinelcore/sentinelcore/internal/remediation"
	"fmt"
	"strings"
)

// SARIF 2.1.0 types — minimal subset for generation.

type sarifLog struct {
	Schema  string     `json:"$schema"`
	Version string     `json:"version"`
	Runs    []sarifRun `json:"runs"`
}

type sarifRun struct {
	Tool        sarifTool         `json:"tool"`
	Results     []sarifResult     `json:"results"`
	Invocations []sarifInvocation `json:"invocations,omitempty"`
}

type sarifTool struct {
	Driver sarifDriver `json:"driver"`
}

type sarifDriver struct {
	Name           string      `json:"name"`
	Version        string      `json:"version"`
	InformationURI string      `json:"informationUri"`
	Rules          []sarifRule `json:"rules"`
}

type sarifRule struct {
	ID               string            `json:"id"`
	Name             string            `json:"name"`
	ShortDescription sarifMsg          `json:"shortDescription"`
	FullDescription  *sarifMsg         `json:"fullDescription,omitempty"`
	Help             *sarifHelp        `json:"help,omitempty"`
	DefaultConfig    sarifDefaultCfg   `json:"defaultConfiguration"`
	Properties       map[string]any    `json:"properties,omitempty"`
}

type sarifMsg struct {
	Text string `json:"text"`
}

type sarifHelp struct {
	Text     string `json:"text"`
	Markdown string `json:"markdown,omitempty"`
}

type sarifDefaultCfg struct {
	Level string `json:"level"`
}

type sarifResult struct {
	RuleID       string              `json:"ruleId"`
	RuleIndex    int                 `json:"ruleIndex"`
	Level        string              `json:"level"`
	Message      sarifMsg            `json:"message"`
	Locations    []sarifLocation     `json:"locations"`
	Fingerprints map[string]string   `json:"fingerprints,omitempty"`
	CodeFlows    []sarifCodeFlow     `json:"codeFlows,omitempty"`
}

type sarifLocation struct {
	Physical *sarifPhysical `json:"physicalLocation,omitempty"`
}

type sarifPhysical struct {
	Artifact sarifArtifact `json:"artifactLocation"`
	Region   *sarifRegion  `json:"region,omitempty"`
}

type sarifArtifact struct {
	URI string `json:"uri"`
}

type sarifRegion struct {
	StartLine int `json:"startLine"`
}

type sarifCodeFlow struct {
	ThreadFlows []sarifThreadFlow `json:"threadFlows"`
}

type sarifThreadFlow struct {
	Locations []sarifTFLocation `json:"locations"`
}

type sarifTFLocation struct {
	Location sarifLocation `json:"location"`
	Kinds    []string      `json:"kinds,omitempty"`
	Message  *sarifMsg     `json:"message,omitempty"`
}

type sarifInvocation struct {
	Success   bool    `json:"executionSuccessful"`
	StartTime *string `json:"startTimeUtc,omitempty"`
	EndTime   *string `json:"endTimeUtc,omitempty"`
}

// FindingSARIF generates a SARIF 2.1.0 log for a single finding.
func FindingSARIF(f FindingData) ([]byte, error) {
	rule := buildRule(f)
	result := buildResult(f, 0)
	log := sarifLog{
		Schema:  "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json",
		Version: "2.1.0",
		Runs: []sarifRun{{
			Tool:    sarifTool{Driver: sarifDriver{Name: "SentinelCore", Version: "1.0.0", InformationURI: "https://sentinelcore.resiliencetech.com.tr", Rules: []sarifRule{rule}}},
			Results: []sarifResult{result},
		}},
	}
	return json.MarshalIndent(log, "", "  ")
}

// ScanSARIF generates a SARIF 2.1.0 log for an entire scan.
func ScanSARIF(d ScanData) ([]byte, error) {
	ruleMap := map[string]int{}
	var rules []sarifRule
	var results []sarifResult

	for _, f := range d.Findings {
		rid := effectiveRuleID(f)
		idx, ok := ruleMap[rid]
		if !ok {
			idx = len(rules)
			ruleMap[rid] = idx
			rules = append(rules, buildRule(f))
		}
		results = append(results, buildResult(f, idx))
	}

	run := sarifRun{
		Tool:    sarifTool{Driver: sarifDriver{Name: "SentinelCore", Version: "1.0.0", InformationURI: "https://sentinelcore.resiliencetech.com.tr", Rules: rules}},
		Results: results,
	}

	if d.StartedAt != nil {
		inv := sarifInvocation{Success: d.Status == "completed"}
		s := d.StartedAt.Format("2006-01-02T15:04:05Z")
		inv.StartTime = &s
		if d.FinishedAt != nil {
			e := d.FinishedAt.Format("2006-01-02T15:04:05Z")
			inv.EndTime = &e
		}
		run.Invocations = []sarifInvocation{inv}
	}

	log := sarifLog{
		Schema:  "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json",
		Version: "2.1.0",
		Runs:    []sarifRun{run},
	}
	return json.MarshalIndent(log, "", "  ")
}

func buildRule(f FindingData) sarifRule {
	rid := effectiveRuleID(f)
	r := sarifRule{
		ID:            rid,
		Name:          rid,
		ShortDescription: sarifMsg{Text: f.Title},
		DefaultConfig: sarifDefaultCfg{Level: sevToLevel(f.Severity)},
	}
	if f.Description != "" {
		r.FullDescription = &sarifMsg{Text: f.Description}
	}
	if f.Remediation != nil {
		r.Help = &sarifHelp{
			Text:     f.Remediation.HowToFix,
			Markdown: buildHelpMD(f),
		}
		props := map[string]any{"tags": []string{f.FindingType, f.Severity}}
		cwe := extractCWE(f.Remediation)
		if len(cwe) > 0 {
			props["cwe"] = cwe
		}
		r.Properties = props
	}
	return r
}

func buildResult(f FindingData, ruleIdx int) sarifResult {
	r := sarifResult{
		RuleID:    effectiveRuleID(f),
		RuleIndex: ruleIdx,
		Level:     sevToLevel(f.Severity),
		Message:   sarifMsg{Text: f.Title},
		Locations: []sarifLocation{buildLoc(f)},
	}
	if f.ID != "" {
		r.Fingerprints = map[string]string{"sentinelcore/v1": f.ID}
	}
	if len(f.TaintPaths) > 1 {
		var locs []sarifTFLocation
		for _, s := range f.TaintPaths {
			locs = append(locs, sarifTFLocation{
				Location: sarifLocation{Physical: &sarifPhysical{
					Artifact: sarifArtifact{URI: s.FilePath},
					Region:   &sarifRegion{StartLine: s.LineStart},
				}},
				Kinds:   []string{s.StepKind},
				Message: &sarifMsg{Text: s.Detail},
			})
		}
		r.CodeFlows = []sarifCodeFlow{{ThreadFlows: []sarifThreadFlow{{Locations: locs}}}}
	}
	return r
}

func buildLoc(f FindingData) sarifLocation {
	uri := f.FilePath
	if uri == "" {
		uri = f.URL
	}
	if uri == "" {
		uri = "unknown"
	}
	line := f.LineStart
	if line < 1 {
		line = 1
	}
	return sarifLocation{Physical: &sarifPhysical{
		Artifact: sarifArtifact{URI: uri},
		Region:   &sarifRegion{StartLine: line},
	}}
}

func buildHelpMD(f FindingData) string {
	if f.Remediation == nil {
		return ""
	}
	r := f.Remediation
	var b strings.Builder
	b.WriteString("## " + r.Title + "\n\n")
	b.WriteString(r.Summary + "\n\n")
	b.WriteString("### How to Fix\n\n")
	b.WriteString(r.HowToFix + "\n")
	if r.SafeExample != "" {
		b.WriteString("\n### Safe Example\n```\n" + r.SafeExample + "\n```\n")
	}
	return b.String()
}

func extractCWE(r *remediation.Pack) []string {
	var cwe []string
	for _, ref := range r.References {
		if strings.HasPrefix(ref.Title, "CWE-") {
			cwe = append(cwe, ref.Title)
		}
	}
	return cwe
}

func effectiveRuleID(f FindingData) string {
	if f.RuleID != "" {
		return f.RuleID
	}
	return fmt.Sprintf("sentinelcore/%s/%s", f.FindingType, f.Severity)
}

func sevToLevel(s string) string {
	switch s {
	case "critical", "high":
		return "error"
	case "medium":
		return "warning"
	default:
		return "note"
	}
}

File: web/features/export/sarif.ts
/**
 * SARIF 2.1.0 export for SentinelCore findings.
 *
 * Spec: https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html
 *
 * Produces JSON that is directly ingestible by:
 * - GitHub Code Scanning (via upload-sarif action)
 * - Azure DevOps Advanced Security
 * - GitLab SAST report ingestion
 * - any SARIF-compatible viewer (VS Code SARIF Viewer, etc.)
 *
 * Design:
 * - Deterministic: same findings always produce the same SARIF.
 * - Safe: no raw secret values in the output.
 * - Standards-aligned: only populates fields the spec defines. Does not
 *   fake unsupported fields.
 */
import type { Finding, Scan } from "@/lib/types";

// --- SARIF type definitions (subset of 2.1.0 spec) ---

interface SarifLog {
  $schema: string;
  version: string;
  runs: SarifRun[];
}

interface SarifRun {
  tool: { driver: SarifToolDriver };
  results: SarifResult[];
  invocations?: SarifInvocation[];
}

interface SarifToolDriver {
  name: string;
  version: string;
  informationUri: string;
  rules: SarifRule[];
}

interface SarifRule {
  id: string;
  name: string;
  shortDescription: { text: string };
  fullDescription?: { text: string };
  help?: { text: string; markdown?: string };
  helpUri?: string;
  defaultConfiguration: { level: SarifLevel };
  properties?: Record<string, unknown>;
}

interface SarifResult {
  ruleId: string;
  ruleIndex: number;
  level: SarifLevel;
  message: { text: string };
  locations: SarifLocation[];
  fingerprints?: Record<string, string>;
  codeFlows?: SarifCodeFlow[];
  properties?: Record<string, unknown>;
}

interface SarifLocation {
  physicalLocation?: {
    artifactLocation: { uri: string };
    region?: { startLine: number; startColumn?: number; endLine?: number };
  };
}

interface SarifCodeFlow {
  threadFlows: SarifThreadFlow[];
}

interface SarifThreadFlow {
  locations: SarifThreadFlowLocation[];
}

interface SarifThreadFlowLocation {
  location: SarifLocation;
  nestingLevel?: number;
  kinds?: string[];
  message?: { text: string };
}

interface SarifInvocation {
  executionSuccessful: boolean;
  startTimeUtc?: string;
  endTimeUtc?: string;
}

type SarifLevel = "error" | "warning" | "note" | "none";

// --- Public API ---

/**
 * Export a single finding as a complete SARIF log.
 */
export function exportFindingSarif(finding: Finding): string {
  const rule = findingToRule(finding);
  const result = findingToResult(finding, 0);
  const log = buildLog([rule], [result]);
  return JSON.stringify(log, null, 2);
}

/**
 * Export an entire scan as a SARIF log with all findings.
 */
export function exportScanSarif(scan: Scan, findings: Finding[]): string {
  // Deduplicate rules by ruleId.
  const ruleMap = new Map<string, SarifRule>();
  const ruleIndex = new Map<string, number>();
  const results: SarifResult[] = [];

  for (const f of findings) {
    const ruleId = effectiveRuleId(f);
    if (!ruleMap.has(ruleId)) {
      ruleMap.set(ruleId, findingToRule(f));
      ruleIndex.set(ruleId, ruleMap.size - 1);
    }
    results.push(findingToResult(f, ruleIndex.get(ruleId)!));
  }

  const log = buildLog(
    Array.from(ruleMap.values()),
    results,
    scan.started_at
      ? [
          {
            executionSuccessful: scan.status === "completed",
            startTimeUtc: scan.started_at,
            endTimeUtc: scan.finished_at,
          },
        ]
      : undefined,
  );
  return JSON.stringify(log, null, 2);
}

// --- Internal helpers ---

function buildLog(
  rules: SarifRule[],
  results: SarifResult[],
  invocations?: SarifInvocation[],
): SarifLog {
  const run: SarifRun = {
    tool: {
      driver: {
        name: "SentinelCore",
        version: "1.0.0",
        informationUri: "https://sentinelcore.resiliencetech.com.tr",
        rules,
      },
    },
    results,
  };
  if (invocations) {
    run.invocations = invocations;
  }
  return {
    $schema:
      "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json",
    version: "2.1.0",
    runs: [run],
  };
}

function findingToRule(f: Finding): SarifRule {
  const ruleId = effectiveRuleId(f);
  const rule: SarifRule = {
    id: ruleId,
    name: ruleId,
    shortDescription: { text: f.title },
    defaultConfiguration: { level: severityToLevel(f.severity) },
  };

  if (f.description) {
    rule.fullDescription = { text: f.description };
  }

  // Remediation as help text.
  if (f.remediation) {
    rule.help = {
      text: f.remediation.how_to_fix,
      markdown: buildHelpMarkdown(f),
    };
  }

  // CWE + OWASP as properties (GitHub Code Scanning reads these).
  const props: Record<string, unknown> = {};
  if (f.remediation?.references) {
    const cwe = f.remediation.references
      .filter((r) => r.title.startsWith("CWE-"))
      .map((r) => r.title);
    if (cwe.length > 0) {
      props["cwe"] = cwe;
    }
    const tags = [f.finding_type];
    if (f.severity) tags.push(f.severity);
    props["tags"] = tags;
  }
  if (Object.keys(props).length > 0) {
    rule.properties = props;
  }

  return rule;
}

function findingToResult(f: Finding, ruleIdx: number): SarifResult {
  const result: SarifResult = {
    ruleId: effectiveRuleId(f),
    ruleIndex: ruleIdx,
    level: severityToLevel(f.severity),
    message: { text: f.title + (f.description ? ". " + firstSentence(f.description) : "") },
    locations: [buildLocation(f)],
  };

  // Stable fingerprint for deduplication across runs.
  if (f.id) {
    result.fingerprints = {
      "sentinelcore/v1": f.id,
    };
  }

  // Taint path as codeFlow.
  if (f.taint_paths && f.taint_paths.length > 1) {
    result.codeFlows = [
      {
        threadFlows: [
          {
            locations: f.taint_paths.map((step) => ({
              location: {
                physicalLocation: {
                  artifactLocation: { uri: step.file_path },
                  region: { startLine: step.line_start },
                },
              },
              kinds: [step.step_kind],
              message: { text: step.detail },
            })),
          },
        ],
      },
    ];
  }

  return result;
}

function buildLocation(f: Finding): SarifLocation {
  if (f.file_path) {
    const loc: SarifLocation = {
      physicalLocation: {
        artifactLocation: { uri: f.file_path },
        region: { startLine: f.line_number || 1 },
      },
    };
    return loc;
  }
  // DAST findings use URL — SARIF doesn't have a great URL location model,
  // but we can use the URI field with the full URL.
  if (f.url) {
    return {
      physicalLocation: {
        artifactLocation: { uri: f.url },
        region: { startLine: 1 },
      },
    };
  }
  return {
    physicalLocation: {
      artifactLocation: { uri: "unknown" },
      region: { startLine: 1 },
    },
  };
}

function buildHelpMarkdown(f: Finding): string {
  const rem = f.remediation;
  if (!rem) return "";
  const parts: string[] = [];
  parts.push(`## ${rem.title}`);
  parts.push("");
  parts.push(rem.summary);
  parts.push("");
  parts.push("### How to Fix");
  parts.push("");
  parts.push(rem.how_to_fix);
  if (rem.safe_example) {
    parts.push("");
    parts.push("### Safe Example");
    parts.push("```");
    parts.push(rem.safe_example);
    parts.push("```");
  }
  return parts.join("\n");
}

function effectiveRuleId(f: Finding): string {
  if (f.rule_id) return f.rule_id;
  // Fallback for findings without a rule_id.
  return `sentinelcore/${f.finding_type}/${f.severity}`;
}

function severityToLevel(severity: string): SarifLevel {
  switch (severity) {
    case "critical":
    case "high":
      return "error";
    case "medium":
      return "warning";
    case "low":
    case "info":
      return "note";
    default:
      return "warning";
  }
}

function firstSentence(text: string): string {
  const match = text.match(/^[^.!?]+[.!?]/);
  return match ? match[0].trim() : text.slice(0, 120);
}
